File: 993. Cousins in Binary Tree - Easy/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        if not root or root.val == x or root.val == y:
            return False
    
        queue = deque([root])
               
        while queue:
            size = len(queue)
            is_x_exist = False
            is_y_exist = False
            for _ in range(size):
                node = queue.popleft()
                if node.val == x:
                    is_x_exist = True
                elif node.val == y:
                    is_y_exist = True
                
                if node.left and node.right:
                    if set([node.left.val, node.right.val]) == set([x, y]):
                        return False
                if node.left: 
                    queue.append(node.left)
                if node.right: 
                    queue.append(node.right)
            if is_x_exist and is_y_exist:
                return True

        return False

File: 993. Cousins in Binary Tree - Easy/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
static auto _ = [] () { ios_base::sync_with_stdio(false); cin.tie(nullptr); return 0; } ();
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        if (!root || root->val == x || root->val == y) {
            return false;
        }
        
        queue < TreeNode* > nodes;
        nodes.push(root);
        
        while (!nodes.empty()) {
            int nodes_to_examine = nodes.size();
            bool found_x = false;
            bool found_y = false;
            for (int i = 0; i < nodes_to_examine; ++i) {
                auto node = nodes.front();
                nodes.pop();
                
                if (node->val == x) {
                    found_x = true;
                }
                if (node->val == y) {
                    found_y = true;
                }
                
                if (node->left && node->right) {
                    if ((node->left->val == x && node->right->val == y) ||
                        (node->left->val == y && node->right->val == x)) {
                        return false;
                    }
                }
                
                if (node->left) {
                    nodes.push(node->left);
                }
                if (node->right) {
                    nodes.push(node->right);
                }
            }
            
            if (found_x && found_y) {
                return true;
            }
        }
        
        return false;
    }
};